src/OpenTelemetry/Metrics/Exemplar/SimpleFixedSizeExemplarReservoir.cs

```diff
@@ -1,140 +1,56 @@
 {
-    private readonly int poolSize;
-    private readonly Random random;
-    private readonly Exemplar[] runningExemplars;
-    private readonly Exemplar[] tempExemplars;
+    private readonly Random random = new();
 
-    private long measurementsSeen;
+    private int measurementsSeen;
 
     public SimpleFixedSizeExemplarReservoir(int poolSize)
+        : base(poolSize)
     {
-        this.poolSize = poolSize;
-        this.runningExemplars = new Exemplar[poolSize];
-        this.tempExemplars = new Exemplar[poolSize];
-        this.measurementsSeen = 0;
-        this.random = new Random();
     }
 
-    public override void Offer(long value, ReadOnlySpan<KeyValuePair<string, object?>> tags, int index = default)
+    public override void Offer(in ExemplarMeasurement<long> measurement)
     {
-        this.Offer(value, tags);
+        this.Offer(in measurement);
     }
 
-    public override void Offer(double value, ReadOnlySpan<KeyValuePair<string, object?>> tags, int index = default)
+    public override void Offer(in ExemplarMeasurement<double> measurement)
     {
-        this.Offer(value, tags);
+        this.Offer(in measurement);
     }
 
-    public override Exemplar[] Collect(ReadOnlyTagCollection actualTags, bool reset)
+    protected override void OnCollected()
     {
-        for (int i = 0; i < this.runningExemplars.Length; i++)
-        {
-            this.tempExemplars[i] = this.runningExemplars[i];
-            if (this.runningExemplars[i].FilteredTags != null)
-            {
-                // TODO: Better data structure to avoid this Linq.
-                // This is doing filtered = alltags - storedtags.
-                // TODO: At this stage, this logic is done inside Reservoir.
-                // Kinda hard for end users who write own reservoirs.
-                // Evaluate if this logic can be moved elsewhere.
-                // TODO: The cost is paid irrespective of whether the
-                // Exporter supports Exemplar or not. One idea is to
-                // defer this until first exporter attempts read.
-                this.tempExemplars[i].FilteredTags = this.runningExemplars[i].FilteredTags!.Except(actualTags.KeyAndValues.ToList()).ToList();
-            }
-
-            if (reset)
-            {
-                this.runningExemplars[i].Timestamp = default;
-            }
-        }
-
         // Reset internal state irrespective of temporality.
         // This ensures incoming measurements have fair chance
         // of making it to the reservoir.
         this.measurementsSeen = 0;
-
-        return this.tempExemplars;
     }
 
-    private void Offer(double value, ReadOnlySpan<KeyValuePair<string, object?>> tags)
+    private void Offer<T>(in ExemplarMeasurement<T> measurement)
+        where T : struct
     {
-        if (this.measurementsSeen < this.poolSize)
+        var measurementNumber = this.measurementsSeen++;
```

Comment:
Increment the counter in the end? This would always skip setting measurement number to `0`.